### ISSUE TITLE
Custom URL matcher: Cannot read property 'parameters' of undefined

### ISSUE THREAD

**Issue by SamVerschueren** [angular/angular#14833]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
I'm experimenting with custom matchers with the routing module. What I want to mimic is a path matching `''` like this.

```js
const routes = [
  {
    path: '',
    component: FooComponent
  }
];
```

So what I use as configuration is the following

```js
const routes = [
  {
    matcher: (url: UrlSegment[]) => {
      return url.length === 0 ? ({consumed: url}) : null;
    },
    component: FooComponent
  }
];
```

If the length of the `UrlSegment` is equal to `0`, return a `{consumed: []}` object, otherwise return `null` indicating that there is no match.

This however results in the following error

> Unhandled Promise rejection: Cannot read property 'parameters' of undefined

I tracked down the bug and the root cause of this can be found in the [`match`](https://github.com/angular/angular/blob/2.4.8/modules/%40angular/router/src/recognize.ts#L168-L186) function of [recognize.ts](https://github.com/angular/angular/blob/2.4.8/modules/%40angular/router/src/recognize.ts). More concretely, on [this](https://github.com/angular/angular/blob/2.4.8/modules/%40angular/router/src/recognize.ts#L183) line.

What that line tries to do is get the last item of consumed segments and get the parameters. But because I don't have any segments, it tries to read the property `parameters` on an `undefined` object. So in my opinion we should add a check first to make sure `res.consumed` is not empty.

I would be happy to do a PR for this!

**Expected behavior**
That we can use a custom matcher to match empty routes.

**Minimal reproduction of the problem with instructions**
The plunkr example can be found [here](http://plnkr.co/edit/rGwrpLHS8m7n1MypQLVQ?p=preview). Run the example and open the console to see the errors.

If you go to `src/app.routing.ts` and change the routes to the following definition, it will work as intendend.

```js
const routes = [
  {
    path: '',
    component: FooComponent
  }
];
```

**What is the motivation / use case for changing the behavior?**
It should also be possible to match empty routes in a custom matcher.

**Please tell us about your environment:**
macOS Sierra, VS Code, yarn

* **Angular version:** 2.4.8

* **Browser:** all
 
* **Language:** TypeScript

* **Node (for AoT issues):** `node --version` =  6.8.0

**Comment by SamVerschueren** [comment]
Also a small side note. While I was exploring this feature, I noticed that it was undocumented. I had to look into the source code to find an example. The [Routes class](https://angular.io/docs/ts/latest/api/router/index/Routes-type-alias.html) on the angular.io website states

> `matcher` defines a custom strategy for path matching and supersedes `path` and `pathMatch`. See [UrlMatcher](https://angular.io/docs/ts/latest/api/router/index/UrlMatcher) for more info.

But the `UrlMatcher` link returns a 404. Is there anything I can do in order to improve the docs? Or should I open a new issue for that.

**Comment by MarkPieszak** [comment]
The docs repo is located here, yea you can make an issue or PR there to let them know about that 404: https://github.com/angular/angular.io

@vsavkin knows all about the router, he might be able to guide us here.

**Comment by SamVerschueren** [comment]
@MarkPieszak Thanks! Created a new issue in the docs repository https://github.com/angular/angular.io/issues/3374

**Comment by Foxandxss** [comment]
@MarkPieszak If the issue is API Related, they belong in here.

**Comment by MarkPieszak** [comment]
Makes sense! My mistake

**Comment by kb3eua** [comment]
I ran into the same issue with "strictNullChecks" enabled.  If the "matcher" property can't allow for a return value of null or undefined in the callback function, then it needs to accept a _consumed_ array of length 0 as a workaround.

**Comment by SamVerschueren** [comment]
Can someone take a look at this issue and PR? It's really a small change...

**Comment by wiikka** [comment]
+1

**Comment by alxhub** [comment]
Closed by b399cb2.
